frontend/javascripts/oxalis/view/action-bar/private_links_view.tsx

```diff
@@ -0,0 +1,347 @@
+    initialData: [],
+    refetchOnWindowFocus: false,
+  });
+}
+
+function useCreateLinkMutation(annotationId: string) {
+  const queryClient = useQueryClient();
+  const mutationKey = ["links", annotationId];
+
+  return useMutation(createPrivateLink, {
+    mutationKey,
+    onSuccess: (newLink) => {
+      queryClient.setQueryData(mutationKey, (oldItems: ZarrPrivateLink[] | undefined) =>
+        (oldItems || []).concat([newLink]),
+      );
+    },
+    onError: (err) => {
+      Toast.error(`Could not create link. ${err}`);
+    },
+  });
+}
+
+function useUpdatePrivateLink(annotationId: string) {
+  const queryClient = useQueryClient();
+  const mutationKey = ["links", annotationId];
+
+  return useMutation(updatePrivateLink, {
+    mutationKey,
+    onMutate: async (updatedLinkItem) => {
+      // Cancel any outgoing refetches (so they don't overwrite our optimistic update)
+      await queryClient.cancelQueries(mutationKey);
+
+      // Snapshot the previous value
+      const previousLinks = queryClient.getQueryData(mutationKey);
+
+      // Optimistically update to the new value
+      queryClient.setQueryData(mutationKey, (oldItems: ZarrPrivateLink[] | undefined) =>
+        (oldItems || []).map((link) => (link.id !== updatedLinkItem.id ? link : updatedLinkItem)),
+      );
+
+      // Return a context object with the snapshotted value
+      return { previousLinks };
+    },
+    // If the mutation fails, use the context returned from onMutate to roll back
+    onError: (err, _updatedLinkItem, context) => {
+      Toast.error(`Could not update link. ${err}`);
+      if (context) {
+        queryClient.setQueryData(mutationKey, context.previousLinks);
+      }
+    },
+  });
+}
+
+function useDeleteLinkMutation(annotationId: string) {
+  const queryClient = useQueryClient();
+
+  const mutationKey = ["links", annotationId];
+
+  return useMutation(deletePrivateLink, {
+    mutationKey,
+    onMutate: async (linkIdToDelete) => {
+      // Cancel any outgoing refetches (so they don't overwrite our optimistic update)
+      await queryClient.cancelQueries(mutationKey);
+
+      // Snapshot the previous value
+      const previousLinks = queryClient.getQueryData(mutationKey);
+
+      // Optimistically update to the new value
+      queryClient.setQueryData(mutationKey, (oldItems: ZarrPrivateLink[] | undefined) =>
+        (oldItems || []).filter((link) => link.id !== linkIdToDelete),
+      );
+
+      // Return a context object with the snapshotted value
+      return { previousLinks };
+    },
+    // If the mutation fails, use the context returned from onMutate to roll back
+    onError: (err, _linkIdToDelete, context) => {
+      Toast.error(`Could not delete link. ${err}`);
+      if (context) {
+        queryClient.setQueryData(mutationKey, context.previousLinks);
+      }
+    },
+  });
+}
+
+function UrlInput({ linkItem }: { linkItem: ZarrPrivateLink }) {
+  const url = `https://.../${linkItem.accessToken}`;
```

Comment:
@leowe can you tell me how this url should look like? :)

Reply:
Good question! The base URI is `${DATASTORE_URI}/annotations/zarr/${accessToken}` However, this will yield a directory listing with only a .zgroup and then subdirectories for the layers. Not all tools will be able to handle the top-level link, I assume, but will need the layer-specific ones. I’m not sure what the best way to handle this would be. @normanrz maybe you have some input?

Reply:
> Not all tools will be able to handle the top-level link, I assume, but will need the layer-specific ones.

Do you know which tools are capable of using the "short" url? One could list these as an example. Maybe providing only the short version also applies some pressure on other software to incorporate support for that...
Alternatively, we could have a dropdown for the copy-button which allows to select the layer one wants to reference. However, it's probably tedious to use (especially if it's unclear which software demands which format). which is why I'd rather avoid that..

Reply:
I don't think there is any tool that can work with "short" urls, sadly. Copy buttons for individual layers would be user-friendly. Perhaps in a nested modal or popover?

Reply:
I pushed a change which opens a dropdown list when clicking the copy button. I also thought about the modal or popover approach, but I think the dropdown is the cleanest option.

Reply:
I forgot to include `/data` in the above URI. It should be `{datastore}/data/annotations/zarr/{tokenValue}`